src/types.ts

```diff
@@ -1,5 +1,48 @@
+import EventEmitter from 'events';
+
 export type PrimitiveValue = null | boolean | string | number;
 
+export interface EventProvider {
+  readonly events: EventEmitter;
+  readonly ready: boolean;
+}
```

Comment:
Providers can optionally implement this (backwards compatible) to support events.

All they must do is mark themselves as ready with the boolean, and call `events.emit(xxx)` to support events.